lib/mailbox.dart

```diff
@@ -87,7 +90,44 @@ class Mailbox {
   ///
   /// If mailbox is empty this will synchronously block until message
   /// is available.
-  Uint8List take() => _mutex.runLocked(() {
+  /// If [timeout] is provided then this will block for at most [timeout].
+  /// If the timeout expires before a message is available then this will
+  /// throw a [TimeoutException].
+  /// The smallest [timeout] is 1 seconds. Anything less than 1 second will
+  /// treated as zero seconds.
+  Uint8List take({Duration? timeout}) {
+    if (timeout != null) {
+      return _takeTimed(timeout);
+    } else {
+      return _take();
+    }
+  }
+
+  Uint8List _takeTimed(Duration timeout) {
+    final start = DateTime.now();
+
+    return _mutex.runLocked(
+      timeout: timeout,
+      () {
+        timeout = timeout - (DateTime.now().difference(start));
```

Comment:
This code needs to be inside the loop because you need to update timeout if you woke up spuriously. 

I suggest to write this approximately like this:

```dart
final timeoutBy = timeout != null ? DateTime.now().add(timeout) : null;

return _mutex.runLocked(timeout: timeout, () {
  while (_mailbox.ref.state != ...) {
    _condVar.wait(_mutex, timeout: timeoutBy?.difference(DateTime.now()));
  }
});
```

I also think that timed and non-timed version can basically be the same. (just pass null timeout to `runLocked` and `wait`).

Reply:
I've modified the code to recalc the remaining time within the loop.

I considered using same method but decided the simplicity of two methods would aid long term maintenance.